### PR TITLE
Tekton hub integration

### PR DESCRIPTION
**Pull request by karthikjeeyar**
**Fixes**: 

https://issues.redhat.com/browse/ODC-6134
https://issues.redhat.com/browse/ODC-6135

**Analysis / Root cause**: 
Integrate tektonHub tasks in the pipeline builder, where users can find/install and update a tekton hub tasks in a namespace.

**Solution Description**: 

1. Pull the tasks from  tekton hub api endpoint and add it to the pipeline task catalog
2. Install / update a tekton hub task to the namespace.
3. Show the installing decorator when the API call is in progress.
4. Added access review checks for the tekton hub CTA button.
5. Update the limited permission flow to have only the installed tasks listed in the modal. 

**Todo**: 

- [ ] Update the Alert message with finalized content (will be provided by UX) 

**Screen shots / Gifs for design review**: 

**Add already Installed Task flow:**

Add_existing_task

**Install task flow:**
installing_task

**Update task flow:**
updating_task

Installing and Delete `Add task` Decorators:
image (14)

**Unit test coverage report**: 

**Test setup:**

```
pipeline-quicksearch-utils
    isSelectedVersionInstalled
      ✓ should return false when the attributes is not present in catalogItem (1ms)
      ✓ should return false when the selected version is not installed
      ✓ should return true when the selected version is installed (1ms)
    isTaskVersionInstalled
      ✓ should return false when the attributes is not present in catalogItem
      ✓ should return false when the installed attribute is not present in catalogItem (1ms)
      ✓ should return true when the installed attribute is set in catalogItem
    isOneVersionInstalled
      ✓ should return false when the attributes is not present in catalogItem (1ms)
      ✓ should return false when the none of the version is installed in the cluster
      ✓ should return true when the one version is installed in the cluster
    isSelectedVersionUpgradable
      ✓ should return false when the attributes is not present in catalogItem (1ms)
      ✓ should return false when the passed version is not upgradable
      ✓ should return false when the selected version is already installed in the cluster
      ✓ should return true when the selected version is upgradable in catalogItem (1ms)
    getTaskCtaType
      ✓ should return the Install button type for the uninstalled task
      ✓ should return the Add button type for the installed version
      ✓ should return the Upgrade button type if selected version is upgradable (1ms)
    isInstalledNamespaceTask
      ✓ should return false if the tekton hub task is passed
      ✓ should return false if the namespace task is not installed through pipeline builder
      ✓ should return true if the namespace task is installed through pipeline builder (1ms)
    isExternalTask
      ✓ should return true for the tekton hub task
      ✓ should return false if clusterTask and namespaceTask is passed
    isTaskSearchable
      ✓ should return true for all the tekton hub tasks are searchable (1ms)
      ✓ should return true for all the cluster and namespace tasks not added through pipeline builder
      ✓ should return false for all the namespace tasks added through pipeline builder
    getSelectedVersionUrl
      ✓ should return null if the attirbutes is not present in the catalogItem (1ms)
      ✓ should return the download url when the item and version id is passed
    findInstalledTask
      ✓ should return undefined the catalogItem is not installed through pipeline builder
      ✓ should return the installed Task resource through pipeline builder (1ms)
 ```
**Browser conformance**: 

- [x] Chrome
- [x] Firefox
- [ ] Safari
- [ ] Edge

cc: @serenamarie125 @bgliwa01 @andrewballantyne

### CODE REVIEW

File: frontend/packages/console-shared/src/components/quick-search/QuickSearchList.tsx
Comment (invincibleJai) [inline frontend/packages/console-shared/src/components/quick-search/QuickSearchList.tsx:99]: nit: cane we have as below
```suggestion
                      {item.icon?.node ?? getIcon(item)}
```

File: frontend/packages/console-shared/src/components/quick-search/QuickSearchList.tsx
Comment (invincibleJai) [inline frontend/packages/console-shared/src/components/quick-search/QuickSearchList.tsx:115]: nit:
```suggestion
                          {item.attributes?.categories?.length > 0 ? (
                            <Label color="blue">{item.attributes.categories[0]}</Label>
                          ) : (
```

and we need to check for `categories[0]`?

File: frontend/packages/pipelines-plugin/src/components/catalog/const.ts
Comment (invincibleJai) [inline frontend/packages/pipelines-plugin/src/components/catalog/const.ts:1]: isn't this endpoint id there within operator, so that we can read and could be dynamic

File: frontend/packages/pipelines-plugin/src/components/catalog/hooks/useTaskCategories.ts
Comment (invincibleJai) [inline frontend/packages/pipelines-plugin/src/components/catalog/hooks/useTaskCategories.ts:6]: do we know type here?

File: frontend/packages/pipelines-plugin/src/components/catalog/hooks/useTaskResources.ts
Comment (invincibleJai) [inline frontend/packages/pipelines-plugin/src/components/catalog/hooks/useTaskResources.ts:5]: this one `useTaskResources` and `useTaskCategories` look similar can we have one hook and control `categories/resources` via props?

File: frontend/packages/pipelines-plugin/src/components/catalog/providers/useTekonHubTasksProvider.tsx
Comment (invincibleJai) [inline frontend/packages/pipelines-plugin/src/components/catalog/providers/useTekonHubTasksProvider.tsx:17]: nit: can we avoid lodash reduce with js one?

File: frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-builder/PipelineBuilderForm.tsx
Comment (invincibleJai) [inline frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-builder/PipelineBuilderForm.tsx:181]: nit:
```suggestion
                    formData.tasks.filter((task) => task.metadata?.installing).length > 0
```

File: frontend/packages/pipelines-plugin/src/components/quicksearch/pipeline-quicksearch-utils.ts
Comment (invincibleJai) [inline frontend/packages/pipelines-plugin/src/components/quicksearch/pipeline-quicksearch-utils.ts:12]: nit: `attributes` look optional shall we add check
```suggestion
  return item.attributes?.installed === selectedVersion;
```

File: frontend/packages/pipelines-plugin/src/components/quicksearch/PipelineQuickSearch.tsx
Comment (invincibleJai) [inline frontend/packages/pipelines-plugin/src/components/quicksearch/PipelineQuickSearch.tsx:79]: no action needed on `then` here only on `catch`?

File: frontend/packages/pipelines-plugin/src/components/quicksearch/PipelineQuickSearchDetails.scss
Comment (invincibleJai) [inline frontend/packages/pipelines-plugin/src/components/quicksearch/PipelineQuickSearchDetails.scss:16]: nit
```suggestion
    margin: var(--pf-global--spacer--md) 0;
```

File: frontend/packages/pipelines-plugin/src/components/catalog/const.ts
Comment (karthikjeeyar) [reply frontend/packages/pipelines-plugin/src/components/catalog/const.ts:1]: Backend support for reading the endpoints is not there yet, so for now we are going with tektonhub API

File: frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-builder/PipelineBuilderForm.tsx
Comment (karthikjeeyar) [reply frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-builder/PipelineBuilderForm.tsx:181]: Here the metadata is optional field, that is managed by us to show the installing decorator.

File: frontend/packages/pipelines-plugin/src/components/quicksearch/PipelineQuickSearch.tsx
Comment (karthikjeeyar) [reply frontend/packages/pipelines-plugin/src/components/quicksearch/PipelineQuickSearch.tsx:79]: Yes, In case of successful task creation, the resources will be created and pulled in automatically by K8swatchers.

File: frontend/packages/console-shared/src/components/quick-search/QuickSearchList.tsx
Comment (karthikjeeyar) [reply frontend/packages/console-shared/src/components/quick-search/QuickSearchList.tsx]: Its just a string array, so the length checks should be enough.

File: frontend/packages/pipelines-plugin/src/components/catalog/hooks/useTaskCategories.ts
Comment (karthikjeeyar) [reply frontend/packages/pipelines-plugin/src/components/catalog/hooks/useTaskCategories.ts]: Added all the tektonHub types.

File: frontend/packages/pipelines-plugin/src/components/catalog/hooks/useTaskResources.ts
Comment (karthikjeeyar) [reply frontend/packages/pipelines-plugin/src/components/catalog/hooks/useTaskResources.ts]: Refactored to one Hook and passing the Url as prop.

**Review by andrewballantyne — CHANGES_REQUESTED** [review]
Also we talked about the storage of the loading tasks, please make those changes as discussed.

File: frontend/packages/pipelines-plugin/src/components/quicksearch/const.ts
Comment (karthikjeeyar) [reply frontend/packages/pipelines-plugin/src/components/quicksearch/const.ts:5]: These are internally used to make decisions in switch cases and to handle the type of the CTA. CTA button text strings are translated aleady.

File: frontend/packages/console-shared/src/components/quick-search/QuickSearchDetails.scss
Comment (karthikjeeyar) [reply frontend/packages/console-shared/src/components/quick-search/QuickSearchDetails.scss]: Updated to have it in pipeline usecase alone.

File: frontend/packages/dev-console/src/components/catalog/catalog-view/CatalogView.tsx
Comment (karthikjeeyar) [reply frontend/packages/dev-console/src/components/catalog/catalog-view/CatalogView.tsx:267]: Still facing issue after adding type to memo. :disappointed:

File: frontend/packages/pipelines-plugin/src/components/catalog/hooks/useApiResponse.ts
Comment (karthikjeeyar) [reply frontend/packages/pipelines-plugin/src/components/catalog/hooks/useApiResponse.ts]: Oops, Updated.

File: frontend/packages/pipelines-plugin/src/components/quicksearch/PiplineQuickSearchDetails.tsx
Comment (karthikjeeyar) [reply frontend/packages/pipelines-plugin/src/components/quicksearch/PiplineQuickSearchDetails.tsx]: not required, removed it.

File: frontend/packages/pipelines-plugin/src/types/pipeline.ts
Comment (andrewballantyne) [inline frontend/packages/pipelines-plugin/src/types/pipeline.ts:45]: This is no longer needed, right?

**Review by andrewballantyne — CHANGES_REQUESTED** [review]
Couple other things... 

Somehow the Sidebar styles got mucked up. I think it was in your previous PR (or another PR that I'm unaware of that touched the Pipeline Builder):
image

Also not sure how it happened... I was clicking around in the modal and I got this:
image
I also got a ton of "Failed to fetch" errors from the `useTekonHubTasksProvider.tsx` file. 

Your implementation also is not what we discussed. I'm not sure if that's because it didn't work or because you misunderstood. I don't want to hold up your PR beyond this sprint, if you can fix the UID issue above and other issues I mentioned, we can merge this as-is.

Seek another review for the LGTM when you get a chance, hopefully we can merge this before EOD tomorrow.

File: frontend/packages/pipelines-plugin/src/components/catalog/providers/useTasksProvider.tsx
Comment (karthikjeeyar) [reply frontend/packages/pipelines-plugin/src/components/catalog/providers/useTasksProvider.tsx:18]: annotations are strings,  could be comma separated, so splitting the string into array.

File: frontend/packages/pipelines-plugin/src/types/pipeline.ts
Comment (karthikjeeyar) [reply frontend/packages/pipelines-plugin/src/types/pipeline.ts:45]: Oops, missed this spot.

**Comment by karthikjeeyar** [comment]
Fixed TaskSidebar alignment issue caused by this removing bootstrap  close styles commit 70a267f801fcc5ced2ac68cf5a18e5cad3e158a8

Not able to reproduce the uid of undefined error, It seems like some flake happened in catalogServiceProvider while loading extensions.

> I also got a ton of "Failed to fetch" errors from the useTekonHubTasksProvider.tsx file.

There were a lot of failures for `/api/check-updates` call, the errors are no longer appearing after building the backeend and re-sourcing it.

File: frontend/packages/dev-console/src/components/catalog/catalog-view/CatalogView.tsx
Comment (karthikjeeyar) [reply frontend/packages/dev-console/src/components/catalog/catalog-view/CatalogView.tsx]: Fixed it.

File: frontend/packages/pipelines-plugin/src/components/catalog/providers/useTekonHubTasksProvider.tsx
Comment (karthikjeeyar) [reply frontend/packages/pipelines-plugin/src/components/catalog/providers/useTekonHubTasksProvider.tsx]: removed.

File: frontend/packages/console-shared/src/components/quick-search/QuickSearchList.tsx
Comment (karthikjeeyar) [reply frontend/packages/console-shared/src/components/quick-search/QuickSearchList.tsx]: updated to have prop in the catalogItem data to render a node.

File: frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-builder/hooks.ts
Comment (invincibleJai) [inline frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-builder/hooks.ts:407]: nit: will  forEach be a better fit here

File: frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-builder/utils.ts
Comment (invincibleJai) [inline frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-builder/utils.ts:323]: do we need this?

File: frontend/packages/pipelines-plugin/src/components/quicksearch/pipeline-quicksearch-utils.ts
Comment (invincibleJai) [inline frontend/packages/pipelines-plugin/src/components/quicksearch/pipeline-quicksearch-utils.ts:20]: nit: can we have it like this?

```suggestion
export const isTaskVersionInstalled = (item: CatalogItem): boolean =>  !!item.attributes?.installed;
```

File: frontend/packages/pipelines-plugin/src/components/quicksearch/pipeline-quicksearch-utils.ts
Comment (invincibleJai) [inline frontend/packages/pipelines-plugin/src/components/quicksearch/pipeline-quicksearch-utils.ts:59]:
```suggestion
    item.data.metadata?.annotations[TektonTaskAnnotation.installedFrom] === TEKTONHUB
```

File: frontend/packages/pipelines-plugin/src/components/quicksearch/pipeline-quicksearch-utils.ts
Comment (invincibleJai) [inline frontend/packages/pipelines-plugin/src/components/quicksearch/pipeline-quicksearch-utils.ts:79]:
```suggestion
  return item.attributes.versions.find((v) => v.id.toString() === version)?.rawURL;
```

File: frontend/packages/pipelines-plugin/src/components/quicksearch/pipeline-quicksearch-utils.ts
Comment (invincibleJai) [inline frontend/packages/pipelines-plugin/src/components/quicksearch/pipeline-quicksearch-utils.ts:89]: same as above

File: frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-builder/utils.ts
Comment (karthikjeeyar) [reply frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-builder/utils.ts]: No, removed.

**Comment by invincibleJai** [comment]
/approve
/label qe-approved

Verified the changes, works as expected cc @andrewballantyne

**Review by andrewballantyne — COMMENTED** [review]
/lgtm

**Comment by karthikjeeyar** [comment]
Adding on behalf of the docs & px team as the Epic already contains the acks.
/label px-approved
/label docs-approved

**Comment by karthikjeeyar** [comment]
Rebased the PR to resolve conflicts.

**Pull request merged**
